File: PYTHON/scripts/run_task2.py
from __future__ import annotations

import argparse
from pathlib import Path
import numpy as np
import matplotlib.pyplot as plt

from src.utils.plot_io import save_plot, ensure_output_dirs
from src.utils.io_paths import infer_run_name


def run(gnss_file: Path, imu_file: Path, run_name: str | None = None) -> None:
    if run_name is None:
        run_name = infer_run_name(gnss_file, imu_file)

    gnss_time = np.genfromtxt(gnss_file, delimiter=",", names=True)["Posix_Time"]
    imu_time = np.genfromtxt(imu_file, usecols=(1,), invalid_raise=False)

    dt_gnss = np.diff(gnss_time)
    dt_imu = np.diff(imu_time)

    fs_gnss = 1.0 / dt_gnss.mean() if dt_gnss.size else float("nan")
    fs_imu = 1.0 / dt_imu.mean() if dt_imu.size else float("nan")

    out_dir = ensure_output_dirs(run_name, "Task_2")

    fig, axes = plt.subplots(1, 3, figsize=(15, 4))
    axes[0].plot(gnss_time[1:], dt_gnss)
    axes[0].set_title("GNSS dt")
    axes[0].set_xlabel("Time [s]")
    axes[0].set_ylabel("dt [s]")

    axes[1].plot(imu_time[1:], dt_imu)
    axes[1].set_title("IMU dt")
    axes[1].set_xlabel("Time [s]")
    axes[1].set_ylabel("dt [s]")

    axes[2].axis("off")
    text = (f"fs_GNSS = {fs_gnss:.2f} Hz\n"
            f"fs_IMU = {fs_imu:.2f} Hz\n"
            f"dt_GNSS mean={dt_gnss.mean():.3f} s\n"
            f"dt_IMU mean={dt_imu.mean():.3f} s")
    axes[2].text(0.1, 0.5, text)

    save_plot(fig, out_dir, "task2_dt_fs")
    plt.close(fig)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Task 2: dt/fs plots")
    parser.add_argument("--gnss", type=Path, required=True)
    parser.add_argument("--imu", type=Path, required=True)
    parser.add_argument("--run", type=str, default=None)
    args = parser.parse_args()
    run(args.gnss, args.imu, args.run)

File: MATLAB/scripts/run_task2.m
function run_task2(gnss_file, imu_file, run_name)
%RUN_TASK2 Sampling statistics and plots
addpath(genpath('..'));
if nargin < 3 || isempty(run_name)
    run_name = infer_run_name(gnss_file, imu_file);
end
Tg = readtable(gnss_file);
Ti = readtable(imu_file);
% assume time columns
wg = Tg.Posix_Time; wi = Ti.Var2;
dt_g = diff(wg); dt_i = diff(wi);
fs_g = 1/mean(dt_g); fs_i = 1/mean(dt_i);
fig = figure('Position',[100 100 900 300]);
subplot(1,3,1); plot(wg(2:end), dt_g); title('GNSS dt'); xlabel('Time [s]'); ylabel('dt [s]');
subplot(1,3,2); plot(wi(2:end), dt_i); title('IMU dt'); xlabel('Time [s]'); ylabel('dt [s]');
subplot(1,3,3); axis off; text(0.1,0.5,{sprintf('fs_{GNSS}=%.2f',fs_g), sprintf('fs_{IMU}=%.2f',fs_i)});
out_dir = fullfile('OUTPUTS', run_name, 'Task_2');
save_plot_png(fig, out_dir, 'task2_dt_fs');
close(fig);
end

function name = infer_run_name(gnss_file, imu_file)
t = regexp(gnss_file,'X\d+','match');
if ~isempty(t), name=t{1}; return; end
t = regexp(imu_file,'X\d+','match');
if ~isempty(t), name=t{1}; else name='RUN'; end
end
